src/marqo/core/inference/inference_models/hugging_face_model.py

```diff
@@ -0,0 +1,249 @@
+    def _load_from_zip_file(self) -> Tuple:
+        """Load the model from a zip file."""
+        zip_file_path = download_model(
+            repo_location=self.model_properties.model_location,
+            url=self.model_properties.url,
+            auth=self.model_auth,
+            download_dir=ModelCache.hf_cache_path
+        )
+
+        model_dir = self.extract_huggingface_archive(zip_file_path)
+        try:
+            model = AutoModel.from_pretrained(model_dir).to(self.device)
+            tokenizer = AutoTokenizer.from_pretrained(model_dir)
+        except (OSError, ValueError, RuntimeError) as e:
+            raise InvalidModelPropertiesError(
+                f"Marqo encountered an error loading the Hugging Face model, modelProperties={self.model_properties}. "
+                f"Please ensure that the provided zip file is valid.\n"
+                f" Original error message = {e}") from e
+        return model, tokenizer
+
+    def _load_pooling_method(self) -> Callable:
+        """Load the pooling method for the model."""
+        if self.model_properties.pooling_method == PoolingMethod.Mean:
+            return self._average_pool_func
+        elif self.model_properties.pooling_method == PoolingMethod.CLS:
+            return self._cls_pool_func
+        else:
+            raise InternalError(f"Invalid pooling method: {self.model_properties.pooling_method}")
+
+    def encode(self, sentence: Union[str, List[str]], normalize=True, **kwargs) -> Union[FloatTensor, np.ndarray]:
+        if isinstance(sentence, str):
+            sentence = [sentence]
+
+        if self._model is None:
+            self.load()
+
+        tokenized_texts = self._tokenizer(
+            sentence,
+            padding=True,
+            truncation=True,
+            max_length=self.model_properties.tokens,
+            return_tensors="pt"
+        ).to(self.device)
+
+        with torch.no_grad():
+            model_output = self._model(**tokenized_texts)
+
+        attention_mask = tokenized_texts['attention_mask']
+
+        embeddings = self._pooling_func(model_output, attention_mask)
+
+        if normalize:
+            embeddings = F.normalize(embeddings, p=2, dim=1)
+
+        return self._convert_output(embeddings)
+
+    def _convert_output(self, output):
+        if self.device == 'cpu':
+            return output.numpy()
+        elif self.device.startswith('cuda'):
+            return output.cpu().numpy()
+
+    @staticmethod
+    def _average_pool_func(model_output, attention_mask):
+        """A pooling function that averages the hidden states of the model."""
+        last_hidden = model_output.last_hidden_state.masked_fill(~attention_mask[..., None].bool(), 0.0)
+        return last_hidden.sum(dim=1) / attention_mask.sum(dim=1)[..., None]
+
+    @staticmethod
+    def _cls_pool_func(model_output, attention_mask=None):
+        """A pooling function that extracts the CLS token from the model."""
+        return model_output[0][:, 0]
```

Comment:
Are these pooling methods common across models? If so, we should consider extract them out, maybe to a PoolingMethod class hierarchy. (does not need to change now)

Reply:
Only used in HF models right now